components/rendering/src/markdown.rs

```diff
@@ -49,6 +49,70 @@ fn is_colocated_asset_link(link: &str) -> bool {
         && !link.starts_with("mailto:")
 }
 
+fn fix_link(link: &str, context: &RenderContext) -> Result<String> {
+    // A few situations here:
+    // - it could be a relative link (starting with `./`)
+    // - it could be a link to a co-located asset
+    // - it could be a normal link
+    // - any of those can be in a header or not: if it's in a header
+    //   we need to append to a string
+    let result = if link.starts_with("./") {
+        match resolve_internal_link(&link, context.permalinks) {
+            Ok(url) => url,
+            Err(_) => {
+                return Err(format!("Relative link {} not found.", link).into());
+            }
+        }
+    } else if is_colocated_asset_link(&link) {
+        format!("{}{}", context.current_page_permalink, link)
+    } else if context.config.check_external_links
+        && !link.starts_with('#')
+        && !link.starts_with("mailto:") {
+        let res = check_url(&link);
+        if res.is_valid() {
+            link.to_string()
+        } else {
+            return Err(
+                format!("Link {} is not valid: {}", link, res.message()).into(),
+            );
+        }
+    } else {
+        link.to_string()
+    };
+    Ok(result)
+}
+
+fn start_tag(temp_header: &mut TempHeader, tag: &Tag) -> bool {
```

Comment:
can we find better name than start_tag/end_tag? it's not really clear what is it doing from the signature